Feature send json

gsabatie opened the pull request.

gsabatie committed:
added router

gsabatie committed:
added router

gsabatie committed:
added VulkanProviderInfos

gsabatie committed:
added VulkanProviderInfos

Commit:
delete useless comment + created better methods to query the data

Commit:
moved inl in proper foldier + disable warning for visuals tudio 2015

Commit:
had deleted the inl from src but forgot to move it in include

Commit:
fix compile error for const reference VulkanInfoProvider::getJSONVulkanInfo()

gsabatie committed:
send json to api

send json to api

gsabatie committed:
Merge remote-tracking branch 'origin/feature-send-json' into feature-send-json

# Conflicts:
#	include/Router.hpp
#	include/VulkanInfoProvider.hpp
#	src/Application.cpp
#	src/Router.cpp
#	src/VulkanInfoProvider.cpp

gsabatie committed:
added put Vulkaninfos

gsabatie committed:
added get vulkan info request

gsabatie committed:
fixed send empty request

gsabatie committed:
Norm

Commit:
fix lugbench compilation about Core::fromInt

Commit:
fix fromToStr to ToStrVec

Quentin01 commented on CMakeLists.txt:
```diff
@@ -96,14 +96,16 @@ include_directories(${LUG_INCLUDE_DIR})
-    src/Router.cpp
+    src/APIClient/Router.cpp
     src/VulkanInfoProvider.cpp
+	src/APIClient/GPURequestor.cpp
```
Indentation using 4 spaces, not tab.

Quentin01 commented on CMakeLists.txt:
```diff
@@ -96,14 +96,16 @@ include_directories(${LUG_INCLUDE_DIR})
-    include/Router.hpp
     include/VulkanInfoProvider.hpp
+    include/APIClient/Router.hpp
+	include/APIClient/GPURequestor.hpp
```
Indentation using 4 spaces, not tab.

Quentin01 commented on CMakeLists.txt:
```diff
@@ -162,3 +164,8 @@ endif()
+
+#add api VERSION
+set(DEFAULT_LUGBENCH_API_VERSION "api/v1")
+lug_set_option(LUGBENCH_API_VERSION ${DEFAULT_LUGBENCH_API_VERSION} STRING "Version of the Lugbench API")
```
You don't have to use a "default" variable here, I was using one for the API URI because the default changes between debug and release, it's not the case for this variable.

Quentin01 replied:
Isn't it better to just store `v1` in this variable? Because the name is `LUGBENCH_API_VERSION` and not `LUGBENCH_API_VERSION_URI` and we will always use the URI `api/{version}`.

Quentin01 commented on CMakeLists.txt:
```diff
@@ -162,3 +164,8 @@ endif()
+#add api VERSION
+set(DEFAULT_LUGBENCH_API_VERSION "api/v1")
+lug_set_option(LUGBENCH_API_VERSION ${DEFAULT_LUGBENCH_API_VERSION} STRING "Version of the Lugbench API")
+add_definitions(/DLUGBENCH_API_VERSION="${LUGBENCH_API_VERSION}")
```
New line at the end of a file

Quentin01 commented on include/APIClient/GPURequestor.hpp:
```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include "restclient-cpp/restclient.h"
```
Use `<>` for your includes.

Quentin01 commented on include/APIClient/GPURequestor.hpp:
```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include "restclient-cpp/restclient.h"
+#include "restclient-cpp/connection.h"
```
Use `<>` for your includes.

Quentin01 commented on include/APIClient/GPURequestor.hpp:
```diff
@@ -0,0 +1,41 @@
+#include "restclient-cpp/connection.h"
+
+#if defined(LUG_SYSTEM_WINDOWS)
+#pragma warning(push)
```
Inside of a `#if` you must indent.

Quentin01 commented on include/APIClient/GPURequestor.hpp:
```diff
@@ -0,0 +1,41 @@
+#endif
+#include <json/json.hpp>
+#if defined(LUG_SYSTEM_WINDOWS)
+#pragma warning(pop)
```
Inside of a `#if` you must indent.

Quentin01 commented on include/APIClient/GPURequestor.hpp:
```diff
@@ -0,0 +1,41 @@
+#pragma warning(pop)
+#endif
+
+#include "Router.hpp"
```
Use `<>` for your includes.

Quentin01 commented on include/APIClient/GPURequestor.hpp:
```diff
@@ -0,0 +1,41 @@
+
+#include "Router.hpp"
+
+using JSONResponse = std::tuple<int, nlohmann::json>;
```
Why these `using` aren't in the namespace `APIClient`?

Quentin01 replied:
Isn't it better to have `using JSONResponse = std::tuple<ResponseCode, nlohmann::json>`?

Quentin01 replied:
It's probably not the good file to put these `using` because you will use them in others classes for interacting with the API.

Quentin01 commented on include/APIClient/GPURequestor.hpp:
```diff
@@ -0,0 +1,41 @@
+using ResponseCode = int;
+
+namespace APIClient {
+class GPURequestor {
```
New line between the `namespace` and the declaration of the `class`.

Quentin01 replied:
I'm not sure for the name of this class, it's redundant to specify `Requestor` because we are already in the namespace `APIClient`, so we know that it's something to do requests on the API.

Maybe `GPU` is just a better name? @Sumenia What do you think of that?

Quentin01 replied:
Isn't it better to use a non-instantiable class?
See the commentary about non-instantiable class in the file `Router.cpp` for the arguments.

Quentin01 commented on include/APIClient/GPURequestor.hpp:
```diff
@@ -0,0 +1,41 @@
+public:
+    GPURequestor() = default;
+
+    GPURequestor(const GPURequestor&) = delete;
```
Where is the constructor by movement?

Quentin01 commented on include/APIClient/GPURequestor.hpp:
```diff
@@ -0,0 +1,41 @@
+
+    ~GPURequestor() = default;
+
+    ResponseCode putVulkanInfo(nlohmann::json vulkanInfosJson);
```
If the API return an error it will probably also send some information about the error in JSON. So, it's probably better to also return a `JSONResponse`.

Quentin01 replied:
I'm not sure, but isn't it better to rename the method `put`? Because for the API it's a GPU, not just Vulkan information.
The same applies for the others methods.

Quentin01 commented on include/APIClient/GPURequestor.hpp:
```diff
@@ -0,0 +1,41 @@
+
+    ResponseCode putVulkanInfo(nlohmann::json vulkanInfosJson);
+    JSONResponse getAllVulkanInfo();
+    JSONResponse getVulkanInfoWithID(uint16_t id);
```
The id isn't a `uint16_t`. Is it really useful to put this method in this PR without testing it? And without having a use for it.

Quentin01 replied:
I think that you need to rename this one `getVulkanInfoWithId` because you are using `Id` for the route and not `ID`.

Quentin01 replied:
I'm not sure, but isn't it better to rename the method `get`?

Quentin01 commented on include/APIClient/GPURequestor.hpp:
```diff
@@ -0,0 +1,41 @@
+    ~GPURequestor() = default;
+
+    ResponseCode putVulkanInfo(nlohmann::json vulkanInfosJson);
+    JSONResponse getAllVulkanInfo();
```
Is it really useful to put this method in this PR without testing it? And without having a use for it.

Quentin01 replied:
I'm not sure, but isn't it better to rename the method `getAll`?

Quentin01 commented on include/APIClient/GPURequestor.hpp:
```diff
@@ -0,0 +1,41 @@
+private:
+    JSONResponse getRequestResponse(std::string request);
+
+    Router _rout;
```
Why are you not naming this property `_router`?

Quentin01 commented on include/APIClient/GPURequestor.hpp:
```diff
@@ -0,0 +1,41 @@
+
+    Router _rout;
+};
+} // APIClient
```
Space after the end of the class declaration

Quentin01 replied:
New line at the end of the file

Quentin01 commented on include/APIClient/Router.hpp:
```diff
@@ -5,24 +5,27 @@
 
 using PutFormData = std::map<std::string, std::string>;
```
Is it useful for the user to have? Or is it just use internally?
You can probably put it in the namespace `APIClient` or in the class `Router` if it's useful for the user of the class.

Quentin01 commented on include/APIClient/GPURequestor.hpp:
```diff
@@ -0,0 +1,41 @@
+    JSONResponse getVulkanInfoWithID(uint16_t id);
+
+private:
+    JSONResponse getRequestResponse(std::string request);
```
Is it really a method for this class? It's probably a method that you want to use everywhere you are doing request to the API.

Quentin01 commented on include/APIClient/Router.hpp:
```diff
@@ -5,24 +5,27 @@
 using PutFormData = std::map<std::string, std::string>;
 
+namespace APIClient {
 class Router {
```
New line between the `namespace` and the `class` declaration.

Quentin01 replied:
I'm not sure why you are using a class instantiable for that. The `Router` doesn't hold any state in the instance, it's probably better to use a non-instantiable class with just `static` methods. (or better, a `namespace`)

Quentin01 commented on include/APIClient/Router.hpp:
```diff
@@ -5,24 +5,27 @@
-    static constexpr const char* baseAPIUri = LUGBENCH_API_URI;
+    std::string getUrlString(Route route, uint16_t id);
 };
+} // APIClient
```
New line at the end of the file.

Quentin01 commented on include/APIClient/Router.hpp:
```diff
@@ -5,24 +5,27 @@
-
-public:
-    static constexpr const char* baseAPIUri = LUGBENCH_API_URI;
+    std::string getUrlString(Route route, uint16_t id);
```
So, you are using a method that takes an `id` even if just one of the three endpoints take one? I think that something is wrong with that but I'm not sure how to resolve this one.

Quentin01 commented on include/VulkanInfoProvider.hpp:
```diff
@@ -3,19 +3,20 @@
 #if defined(LUG_SYSTEM_WINDOWS)
-    #pragma warning(push)
-    #pragma warning(disable : 4003)
+#pragma warning(push)
```
You need indentation inside a `#if`

Quentin01 commented on include/VulkanInfoProvider.hpp:
```diff
@@ -3,19 +3,20 @@
 #include <json/json.hpp>
 #if defined(LUG_SYSTEM_WINDOWS)
-    #pragma warning(pop)
+#pragma warning(pop)
```
You need indentation inside a `#if`

Quentin01 commented on include/VulkanInfoProvider.hpp:
```diff
@@ -3,19 +3,20 @@
+#pragma warning(pop)
 #endif
 
 class VulkanInfoProvider {
```
So, why do you need an instantiable class for just one method? Like I have said for `GPURequestor` and `Router` it's probably better to use a non-instantiable class.

Quentin01 replied:
For the API, it's a GPU entry and not just vulkan information (and later we will have an entry for the computer with some information like the RAM, the CPU, etc and each computer will have one or more GPUs). So, can we rename this class `GPUInfoProvider` for example?

Quentin01 commented on include/VulkanInfoProvider.hpp:
```diff
@@ -3,19 +3,20 @@
 class VulkanInfoProvider {
 public:
-    VulkanInfoProvider(lug::Graphics::Vulkan::InstanceInfo& instanceInfo);
+    VulkanInfoProvider(lug::Graphics::Vulkan::PhysicalDeviceInfo *physicalDeviceInfo);
```
You always need the `PhysicalDeviceInfo`, you don't want `nullptr`. So use a reference instead of a pointer.

Quentin01 commented on include/VulkanInfoProvider.inl:
```diff
@@ -0,0 +1,3 @@
+inline const nlohmann::json& VulkanInfoProvider::getJSONVulkanInfo() const {
```
I think that this file is not used anymore, it's not included in `VulkanInfoProvider.hpp` and the property `_json` doesn't exist.

Quentin01 commented on src/APIClient/GPURequestor.cpp:
```diff
@@ -0,0 +1,25 @@
+#include "APIClient/GPURequestor.hpp"
```
Use `<>` for your includes

Quentin01 commented on src/APIClient/GPURequestor.cpp:
```diff
@@ -0,0 +1,25 @@
+        json = nlohmann::json(r.body);
+    }
+    return std::make_tuple(r.code, json);
+}
```
New line at the end of the file

Quentin01 commented on src/APIClient/GPURequestor.cpp:
```diff
@@ -0,0 +1,25 @@
+#include "APIClient/GPURequestor.hpp"
+
+int APIClient::GPURequestor::putVulkanInfo(nlohmann::json vulkanInfosJson) {
```
Why are you returning an `int`? It's not the same type as the prototype in the header file.

Quentin01 commented on src/APIClient/GPURequestor.cpp:
```diff
@@ -0,0 +1,25 @@
+#include "APIClient/GPURequestor.hpp"
+
+int APIClient::GPURequestor::putVulkanInfo(nlohmann::json vulkanInfosJson) {
+    std::cout << vulkanInfosJson.dump() << std::endl;
```
Do you really want to display the JSON here? I don't really care if you display or not the JSON, but it's probably better to do it in another place.

Quentin01 commented on src/APIClient/GPURequestor.cpp:
```diff
@@ -0,0 +1,25 @@
+}
+
+JSONResponse APIClient::GPURequestor::getAllVulkanInfo() {
+    return	getRequestResponse(_rout.getUrlString(Router::Route::getAllVulkanInfo, 0));
```
Why do you have so many spaces between `return` and the content to return?

Quentin01 commented on src/APIClient/GPURequestor.cpp:
```diff
@@ -0,0 +1,25 @@
+}
+
+JSONResponse APIClient::GPURequestor::getVulkanInfoWithID(uint16_t id) {
+    return	getRequestResponse(_rout.getUrlString(Router::Route::getVulkanInfoWithId, id));
```
Why do you have so many spaces between `return` and the content to return?

Quentin01 commented on src/APIClient/GPURequestor.cpp:
```diff
@@ -0,0 +1,25 @@
+int APIClient::GPURequestor::putVulkanInfo(nlohmann::json vulkanInfosJson) {
+    std::cout << vulkanInfosJson.dump() << std::endl;
+    RestClient::Response r = RestClient::put(_rout.getUrlString(Router::Route::putVulkanInfo, 0), "application/json;charset=utf-8", vulkanInfosJson.dump(0));
+    return r.code;
```
If you want to just return the code, you don't need the temporary variable `r`.

Quentin01 commented on src/APIClient/GPURequestor.cpp:
```diff
@@ -0,0 +1,25 @@
+    nlohmann::json json;
+    RestClient::Response r = RestClient::get(request);
+
+    if (r.code >= 300 && r.code < 400) {
```
The API will probably returns a JSON in case of error (to return some information about the error) or success (to return the value of the request), so why do you bother checking that?

Quentin01 commented on src/APIClient/Router.cpp:
```diff
@@ -0,0 +1,23 @@
+#include "APIClient/Router.hpp"
```
Use `<>` for your includes

Quentin01 commented on src/APIClient/Router.cpp:
```diff
@@ -0,0 +1,23 @@
+constexpr const char* APIClient::Router::baseAPIUri;
+
+std::string APIClient::Router::getUrlString(Route route, uint16_t id = 0) {
+    std::string	URLString;
```
Why do you have so many spaces between the type and the name of your variable?

Quentin01 replied:
Why do you need this variable? You can just return in the switch, it's better.

Quentin01 commented on src/Application.cpp:
```diff
@@ -1,13 +1,18 @@
-#include <lug/Graphics/Vulkan/API/RTTI/Enum.hpp>
 
-Application::Application() : lug::Core::Application::Application{ {"hello", {0, 1, 0}} } {
+#include "restclient-cpp/restclient.h"
```
Use `<>` for your includes

Quentin01 commented on src/Application.cpp:
```diff
@@ -24,8 +29,8 @@ Application::Application() : lug::Core::Application::Application{ {"hello", {0,
 
-            0.0f,                                           // minDepth
-            1.0f                                            // maxDepth
+        0.0f,                                           // minDepth
```
Why do you have remove the indentation?

Quentin01 commented on src/Application.cpp:
```diff
@@ -44,10 +49,7 @@ Application::Application() : lug::Core::Application::Application{ {"hello", {0,
-    for (auto& queue : vkRender->getQueues()) {
-        queue.waitIdle();
-    }
+	vkRender->getDevice().waitIdle();
```
Use 4 spaces instead of a tab for indentation.

Quentin01 commented on src/Application.cpp:
```diff
@@ -58,11 +60,12 @@ bool Application::init(int argc, char* argv[]) {
-            LUG_LOG.warn("Can't initialize the engine for the device {}", choosedDevice.properties.deviceName);
-        }
+    auto& choosedDevice = vkRender->getPhysicalDeviceInfos().front();
+    /*   for (auto& choosedDevice : vkRender->getPhysicalDeviceInfos()) {*/
```
You probably don't want to push that.

Quentin01 commented on src/VulkanInfoProvider.cpp:
```diff
@@ -1,11 +1,293 @@
+        );
+    }
+
+    //    for (auto surface : _physicalDeviceInfo->swapchain)
```
Remove the comments

Quentin01 commented on src/VulkanInfoProvider.cpp:
```diff
@@ -1,11 +1,293 @@
+
+    //    for (auto surface : _physicalDeviceInfo->swapchain)
+    //  {
+    json["zSwapChainInfo"].push_back(
```
Rename this one `"swapchain"`?

Quentin01 commented on src/VulkanInfoProvider.cpp:
```diff
@@ -1,11 +1,293 @@
+
+    }
+    );
+    //
```
Remove the comment

Quentin01 commented on src/VulkanInfoProvider.cpp:
```diff
@@ -1,11 +1,293 @@
+        {"minImageCount", _physicalDeviceInfo->swapchain.capabilities.minImageCount},
+        {"maxImageCount", _physicalDeviceInfo->swapchain.capabilities.maxImageCount},
+        {"currentExtent",
+        {
```
The indentation is wrong, you need to put the bracket at the end of the previous line.

Quentin01 commented on src/VulkanInfoProvider.cpp:
```diff
@@ -1,11 +1,293 @@
+        }
+        },
+        {"minImageExtent",
+        {
```
The indentation is wrong, you need to put the bracket at the end of the previous line.

Quentin01 commented on src/VulkanInfoProvider.cpp:
```diff
@@ -1,11 +1,293 @@
+        }
+        },
+        {"maxImageExtent",
+        {
```
The indentation is wrong, you need to put the bracket at the end of the previous line.

Quentin01 commented on src/VulkanInfoProvider.cpp:
```diff
@@ -1,11 +1,293 @@
+        },
+
+        {"maxImageArrayLayers", _physicalDeviceInfo->swapchain.capabilities.maxImageArrayLayers},
+
```
Why do you have a space here?

Quentin01 commented on src/VulkanInfoProvider.cpp:
```diff
@@ -1,11 +1,293 @@
+            {"height", _physicalDeviceInfo->swapchain.capabilities.maxImageExtent.height}
+        }
+        },
+
```
Why do you have a space here?

Quentin01 commented on src/main.cpp:
```diff
@@ -1,24 +1,24 @@
 #if defined(LUG_SYSTEM_WINDOWS)
-    #pragma warning(push)
-    #pragma warning(disable : 4003)
+#pragma warning(push)
```
Indentation inside a `#if`

Quentin01 commented on src/main.cpp:
```diff
@@ -1,24 +1,24 @@
 #include <json/json.hpp>
 #if defined(LUG_SYSTEM_WINDOWS)
-    #pragma warning(pop)
+#pragma warning(pop)
```
Indentation inside a `#if`

Quentin01 commented on src/main.cpp:
```diff
@@ -1,24 +1,24 @@
 #include <lug/System/Logger/Logger.hpp>
 #if defined(LUG_SYSTEM_ANDROID)
-    #include <lug/System/Logger/LogCatHandler.hpp>
+#include <lug/System/Logger/LogCatHandler.hpp>
```
Indentation inside a `#if`

Quentin01 commented on src/main.cpp:
```diff
@@ -1,24 +1,24 @@
+#include <lug/System/Logger/LogCatHandler.hpp>
 #else
-    #include <lug/System/Logger/OstreamHandler.hpp>
+#include <lug/System/Logger/OstreamHandler.hpp>
```
Indentation inside a `#if`

Quentin01 commented on src/main.cpp:
```diff
@@ -1,24 +1,24 @@
 
 #include "Application.hpp"
-#include "Router.hpp"
+#include "APIClient/Router.hpp"
```
Use `<>` for your includes

gsabatie replied:
fixed

Nokitoo commented on CMakeLists.txt:
```diff
@@ -96,14 +96,16 @@ include_directories(${LUG_INCLUDE_DIR})
-    src/Router.cpp
+    src/APIClient/Router.cpp
     src/VulkanInfoProvider.cpp
+    src/APIClient/GPURequestor.cpp
```
It has to be sorted in alphabetic order

Nokitoo commented on CMakeLists.txt:
```diff
@@ -96,14 +96,16 @@ include_directories(${LUG_INCLUDE_DIR})
-    include/Router.hpp
     include/VulkanInfoProvider.hpp
+    include/APIClient/Router.hpp
+    include/APIClient/GPURequestor.hpp
```
It has to be sorted in alphabetic order

Nokitoo commented on CMakeLists.txt:
```diff
@@ -162,3 +164,8 @@ endif()
 lug_set_option(LUGBENCH_API_URI ${DEFAULT_LUGBENCH_API_URI} STRING "URI of the Lugbench API")
 add_definitions(/DLUGBENCH_API_URI="${LUGBENCH_API_URI}")
+
+#add api VERSION
```
Must have a space after `#`

Nokitoo commented on include/APIClient/Router.hpp:
```diff
@@ -5,24 +5,27 @@
+
+    static constexpr const char* baseAPIUri = LUGBENCH_API_URI;
+    static constexpr const char* apiVersion = LUGBENCH_API_VERSION;
+    static constexpr const char* apiVersionUri = "api/";
```
The class enum/typedef/static variables and class methods should have a different block like this:
```
public:
    enum class Route : uint8_t {
        // ....
    };
    // ....

public:
    Route() = default;
    // ....
```

gsabatie replied:
Router is not a class anymore, it is just a namespace

Nokitoo commented on include/VulkanInfoProvider.hpp:
```diff
@@ -3,19 +3,20 @@
 #include <lug/Config.hpp>
 #include <lug/Graphics/Renderer.hpp>
 #include <lug/Graphics/Vulkan/Renderer.hpp>
+#include <lug/Graphics/Vulkan/API/RTTI/Enum.hpp>
```
Includes have to be sorted in alphabetic order

Nokitoo commented on src/APIClient/GPURequestor.cpp:
```diff
@@ -0,0 +1,25 @@
+    return	getRequestResponse(_rout.getUrlString(Router::Route::getVulkanInfoWithId, id));
+}
+
+JSONResponse APIClient::GPURequestor::getRequestResponse(std::string request) {
```
Why don't you pass `request` as `const std::string&` ?

Quentin01 commented on CMakeLists.txt:
```diff
@@ -94,16 +94,21 @@ find_package(LUG REQUIRED core graphics system window math)
-    include/Router.hpp
-    include/VulkanInfoProvider.hpp
+    include/GPUInfoProvider.hpp
+   
```
Remove this new line

Quentin01 commented on CMakeLists.txt:
```diff
@@ -94,16 +94,21 @@ find_package(LUG REQUIRED core graphics system window math)
 
 set(INC
+    include/APIClient/GPU.hpp
+    include/APIClient/Common.hpp
```
Alphabetic order

Quentin01 commented on CMakeLists.txt:
```diff
@@ -94,16 +94,21 @@ find_package(LUG REQUIRED core graphics system window math)
 
 set(SRC
+    src/APIClient/GPU.cpp
+    src/APIClient/Common.cpp
```
Alphabetic order

Quentin01 commented on include/APIClient/Common.hpp:
```diff
@@ -0,0 +1,21 @@
+using ResponseCode = int;
+using JSONResponse = std::tuple<ResponseCode, nlohmann::json>;
+
+JSONResponse getRequestResponse(std::string request);
```
You need to take a `const std::string&`

Quentin01 commented on include/APIClient/GPU.hpp:
```diff
@@ -0,0 +1,31 @@
+
+class GPU {
+
+    using ResponseCode = int;
```
It's already defined in `APIClient/Common.hpp`

Quentin01 commented on include/APIClient/Router.hpp:
```diff
@@ -0,0 +1,22 @@
+#include <map>
+
+namespace APIClient {
+
```
No need to put a new line between two namespaces

Quentin01 commented on include/APIClient/Router.hpp:
```diff
@@ -0,0 +1,22 @@
+    static constexpr const char* apiVersion = LUGBENCH_API_VERSION;
+    static constexpr const char* apiVersionUri = "api/";
+
+    std::string getUrlString(Route route, std::string id);
```
You need to take a `const std::string&` for the `id`.

Quentin01 replied:
So, is it really okay to always take an id even if two of the three URI don't take an id?

Quentin01 commented on include/APIClient/Router.hpp:
```diff
@@ -0,0 +1,22 @@
+    static constexpr const char* apiVersionUri = "api/";
+
+    std::string getUrlString(Route route, std::string id);
+};
```
No `;` after the declaration of a namespace
Put a comment with the name of the namespace.

Quentin01 commented on include/GPUInfoProvider.hpp:
```diff
@@ -0,0 +1,20 @@
+#endif
+
+namespace GPUInfoProvider {
+
```
No need to put a new line between the namespace and the prototype of the function

Quentin01 commented on include/GPUInfoProvider.hpp:
```diff
@@ -0,0 +1,20 @@
+
+namespace GPUInfoProvider {
+
+    nlohmann::json getJSONVulkanInfo(lug::Graphics::Vulkan::PhysicalDeviceInfo &physicalDeviceInfo);
```
Maybe rename this function `get`? You are already in the namespace `GPUInfoProvider` so it's obvious what `get` is doing.

Quentin01 replied:
You need to take a `const lug::Graphics::Vulkan::PhysicalDeviceInfo&`.
And not space between `&` and the type.

Quentin01 commented on include/APIClient/GPU.hpp:
```diff
@@ -0,0 +1,31 @@
+
+    ~GPU() = default;
+
+    JSONResponse put(nlohmann::json vulkanInfosJson);
```
You need to take a `const nlohmann::json&`.
And you can probably rename the variable (the same in the .cpp) `gpuInfo`

Quentin01 commented on include/APIClient/GPU.hpp:
```diff
@@ -0,0 +1,31 @@
+
+    JSONResponse put(nlohmann::json vulkanInfosJson);
+    JSONResponse getAll();
+    JSONResponse get(std::string id);
```
You need to take a `const std::string&`.

Quentin01 commented on include/GPUInfoProvider.hpp:
```diff
@@ -0,0 +1,20 @@
+namespace GPUInfoProvider {
+
+    nlohmann::json getJSONVulkanInfo(lug::Graphics::Vulkan::PhysicalDeviceInfo &physicalDeviceInfo);
+};
```
No `;` after a namespace declaration.
Put a comment with the name of the namespace.

Quentin01 commented on src/APIClient/GPU.cpp:
```diff
@@ -0,0 +1,16 @@
+    nlohmann::json reponseJson;
+    RestClient::Response r = RestClient::put(APIClient::Router::getUrlString(Router::Route::put, 0), "application/json;charset=utf-8", vulkanInfosJson.dump(0));
+    reponseJson = nlohmann::json(r.body);
+    return std::make_tuple(r.code, reponseJson);
```
Why aren't you using `getRequestResponse`?

gsabatie replied:
because it's a put request and not a get request

Quentin01 replied:
And? It's the same, you have a code and a JSON as the body.

Quentin01 commented on src/APIClient/Router.cpp:
```diff
@@ -0,0 +1,18 @@
+#include <APIClient/Router.hpp>
+
+std::string APIClient::Router::getUrlString(APIClient::Router::Route route, std::string id = "") {
```
A default value for a parameter must be present in the header file, not in the source file.

Quentin01 commented on src/APIClient/Router.cpp:
```diff
@@ -0,0 +1,18 @@
+        case APIClient::Router::Route::put:
+            return std::string(baseAPIUri) + std::string(apiVersionUri) + std::string(apiVersion) + "/gpus";
+#if defined(_WIN32 )
+        default:
```
Don't do that, prefer to add a `return` at the end of the function. It's safer.

Quentin01 commented on src/Application.cpp:
```diff
@@ -1,13 +1,18 @@
-Application::Application() : lug::Core::Application::Application{ {"hello", {0, 1, 0}} } {
+#include <restclient-cpp/restclient.h>
+
+#include <Application.hpp>
```
Alphabetic order

Quentin01 commented on include/APIClient/Router.hpp:
```diff
@@ -0,0 +1,22 @@
+#pragma once
+
+#include <string>
```
Alphabetic order

Quentin01 commented on src/APIClient/Common.cpp:
```diff
@@ -0,0 +1,9 @@
+#include <restclient-cpp/restclient.h>
+
+#include <APIClient/Common.hpp>
```
The guidelines says that you need to put this one first, then a new line, then the others.

Quentin01 commented on include/APIClient/Router.hpp:
```diff
@@ -0,0 +1,22 @@
+namespace Router {
+
+    enum class Route : uint8_t {
+        getAll,
```
You have changed that but it's not obvious what you are doing on this route.
Change it to `getAllGpus`.

Quentin01 commented on include/APIClient/Router.hpp:
```diff
@@ -0,0 +1,22 @@
+
+    enum class Route : uint8_t {
+        getAll,
+        getVulkanInfoWithId,
```
You have changed that but it's not obvious what you are doing on this route.
Change it to `getGpu`.

Quentin01 commented on include/APIClient/Router.hpp:
```diff
@@ -0,0 +1,22 @@
+    enum class Route : uint8_t {
+        getAll,
+        getVulkanInfoWithId,
+        put
```
You have changed that but it's not obvious what you are doing on this route.
Change it to `putGpu`.

Quentin01 commented on src/Application.cpp:
```diff
@@ -24,8 +29,8 @@ Application::Application() : lug::Core::Application::Application{ {"hello", {0,
 
-            0.0f,                                           // minDepth
-            1.0f                                            // maxDepth
+        0.0f,                                               // minDepth
```
Why did you change the indentation?

Quentin01 commented on src/Application.cpp:
```diff
@@ -24,8 +29,8 @@ Application::Application() : lug::Core::Application::Application{ {"hello", {0,
-            0.0f,                                           // minDepth
-            1.0f                                            // maxDepth
+        0.0f,                                               // minDepth
+        1.0f                                                // maxDepth
```
Why did you change the indentation?

Quentin01 commented on src/Application.cpp:
```diff
@@ -78,295 +79,13 @@ bool Application::initDevice(lug::Graphics::Vulkan::PhysicalDeviceInfo* choosedD
     }
 
     lug::Graphics::Vulkan::PhysicalDeviceInfo *physicalDeviceInfo = vkRender->getPhysicalDeviceInfo();
```
Use a `const`

Quentin01 commented on src/Application.cpp:
```diff
@@ -63,7 +65,6 @@ bool Application::init(int argc, char* argv[]) {
             LUG_LOG.warn("Can't initialize the engine for the device {}", choosedDevice.properties.deviceName);
         }
     }
-
```
Why did you remove this new line?

Quentin01 commented on src/Application.cpp:
```diff
@@ -78,295 +79,13 @@ bool Application::initDevice(lug::Graphics::Vulkan::PhysicalDeviceInfo* choosedD
+    std::cout << json.dump() << std::endl;
+    APIClient::GPU req;
+    req.put(json);
     return (true);
```
Do not use parentheses around `return` value.

Quentin01 commented on src/Application.cpp:
```diff
@@ -78,295 +79,13 @@ bool Application::initDevice(lug::Graphics::Vulkan::PhysicalDeviceInfo* choosedD
-
-    std::cout << json.dump(4) << std::endl;
-
+    nlohmann::json json = GPUInfoProvider::getJSONVulkanInfo(*physicalDeviceInfo);
```
You can use a `const`

Quentin01 commented on src/GPUInfoProvider.cpp:
```diff
@@ -0,0 +1,284 @@
+        {"currentExtent", {
+            {"width", physicalDeviceInfo.swapchain.capabilities.currentExtent.width},
+            {"height", physicalDeviceInfo.swapchain.capabilities.currentExtent.height}
+        }
```
Because you are opening two brackets on the same line, you need the two closing ones on the same line too.

Quentin01 commented on src/GPUInfoProvider.cpp:
```diff
@@ -0,0 +1,284 @@
+        {"minImageExtent",{
+            {"width", physicalDeviceInfo.swapchain.capabilities.minImageExtent.width},
+            {"height", physicalDeviceInfo.swapchain.capabilities.minImageExtent.height}
+        }
```
Because you are opening two brackets on the same line, you need the two closing ones on the same line too.

Quentin01 commented on src/GPUInfoProvider.cpp:
```diff
@@ -0,0 +1,284 @@
+        {"maxImageExtent", {
+            {"width", physicalDeviceInfo.swapchain.capabilities.maxImageExtent.width},
+            {"height", physicalDeviceInfo.swapchain.capabilities.maxImageExtent.height}
+        }
```
Because you are opening two brackets on the same line, you need the two closing ones on the same line too.

Quentin01 commented on src/GPUInfoProvider.cpp:
```diff
@@ -0,0 +1,284 @@
+            {"height", physicalDeviceInfo.swapchain.capabilities.maxImageExtent.height}
+        }
+        },
+
```
Remove this empty line

Quentin01 commented on src/GPUInfoProvider.cpp:
```diff
@@ -0,0 +1,284 @@
+        },
+
+        {"maxImageArrayLayers", physicalDeviceInfo.swapchain.capabilities.maxImageArrayLayers},
+
```
Remove this empty line

Quentin01 commented on src/GPUInfoProvider.cpp:
```diff
@@ -0,0 +1,284 @@
+
+    }
+    );
+    return json;
```
Maybe add a new line between this one and the previous one.

Nokitoo commented on include/APIClient/GPU.hpp:
```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include <restclient-cpp/restclient.h>
+#include <restclient-cpp/connection.h>
```
Alphabetic order

Nokitoo commented on include/APIClient/Common.hpp:
```diff
@@ -0,0 +1,21 @@
+#pragma once
+
+#include <tuple>
+#include <string>
```
Alphabetic order

Quentin01 commented on include/APIClient/GPU.hpp:
```diff
@@ -0,0 +1,20 @@
+
+namespace GPU {
+
+    using ResponseCode = int;
```
It's already defined in `APIClient/Common.hpp`

Quentin01 commented on include/APIClient/GPU.hpp:
```diff
@@ -0,0 +1,20 @@
+namespace APIClient {
+
+namespace GPU {
+
```
Remove this empty line

Quentin01 commented on include/APIClient/GPU.hpp:
```diff
@@ -0,0 +1,20 @@
+    JSONResponse put(nlohmann::json vulkanInfosJson);
+    JSONResponse getAll();
+    JSONResponse get(std::string id);
+};
```
Don't put a `;` after a namespace declaration.
Put a comment with the name of the namespace.

Quentin01 commented on include/APIClient/GPU.hpp:
```diff
@@ -0,0 +1,20 @@
+    using ResponseCode = int;
+    using JSONResponse = std::tuple<ResponseCode, nlohmann::json>;
+
+    JSONResponse put(nlohmann::json vulkanInfosJson);
```
Take a `const nlohmann::json&`

Quentin01 commented on include/APIClient/GPU.hpp:
```diff
@@ -0,0 +1,20 @@
+
+    JSONResponse put(nlohmann::json vulkanInfosJson);
+    JSONResponse getAll();
+    JSONResponse get(std::string id);
```
Take a `const std::string&`

Quentin01 commented on include/APIClient/GPU.hpp:
```diff
@@ -0,0 +1,20 @@
+#include <APIClient/Router.hpp>
+
+namespace APIClient {
+
```
No need to put a new line between two namespaces

Quentin01 commented on include/APIClient/GPU.hpp:
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+#include <restclient-cpp/connection.h>
```
Don't include this file, you don't need it in the header.

Quentin01 commented on include/APIClient/GPU.hpp:
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+#include <restclient-cpp/connection.h>
+#include <restclient-cpp/restclient.h>
```
Don't include this file, you don't need it in the header.

Quentin01 commented on include/APIClient/GPU.hpp:
```diff
@@ -0,0 +1,17 @@
+#include <restclient-cpp/restclient.h>
+
+#include <APIClient/Common.hpp>
+#include <APIClient/Router.hpp>
```
Don't include this file, you don't need it in the header.

Quentin01 commented on include/APIClient/Router.hpp:
```diff
@@ -0,0 +1,22 @@
+#pragma once
+
+#include <map>
```
Don't include this file, you don't need it in the header.

Quentin01 commented on include/GPUInfoProvider.hpp:
```diff
@@ -0,0 +1,21 @@
+#pragma once
+
+#include <lug/Config.hpp>
+#include <lug/Graphics/Renderer.hpp>
```
Don't include this file, you don't need it in the header.

Quentin01 commented on include/GPUInfoProvider.hpp:
```diff
@@ -0,0 +1,21 @@
+
+#include <lug/Config.hpp>
+#include <lug/Graphics/Renderer.hpp>
+#include <lug/Graphics/Vulkan/API/RTTI/Enum.hpp>
```
Don't include this file, you don't need it in the header.

Quentin01 commented on include/GPUInfoProvider.hpp:
```diff
@@ -0,0 +1,21 @@
+#include <lug/Config.hpp>
+#include <lug/Graphics/Renderer.hpp>
+#include <lug/Graphics/Vulkan/API/RTTI/Enum.hpp>
+#include <lug/Graphics/Vulkan/Renderer.hpp>
```
You probably want to include `lug/Graphics/Vulkan/Vulkan.hpp` instead

Quentin01 commented on include/APIClient/GPU.hpp:
```diff
@@ -0,0 +1,17 @@
+namespace APIClient {
+namespace GPU {
+
+JSONResponse putGpu(const nlohmann::json& vulkanInfosJson);
```
Is it necessary to repeat `Gpu` in the name?

Quentin01 commented on include/APIClient/GPU.hpp:
```diff
@@ -0,0 +1,17 @@
+namespace GPU {
+
+JSONResponse putGpu(const nlohmann::json& vulkanInfosJson);
+JSONResponse getAllGpus();
```
Is it necessary to repeat `Gpu` in the name?

Quentin01 commented on include/APIClient/Common.hpp:
```diff
@@ -0,0 +1,22 @@
+#include <string>
+#include <tuple>
+
+#if defined(LUG_SYSTEM_WINDOWS)
```
To have `LUG_SYSTEM_WINDOWS` you probably want to include `lug/Config.hpp`

Quentin01 commented on src/APIClient/GPU.cpp:
```diff
@@ -0,0 +1,16 @@
+    nlohmann::json reponseJson;
+    RestClient::Response r = RestClient::put(APIClient::Router::getUrlString(Router::Route::putGpu), "application/json;charset=utf-8", vulkanInfosJson.dump());
+    reponseJson = nlohmann::json(r.body);
+    return std::make_tuple(r.code, reponseJson);
```
So, you have `getRequestResponse`, why not doing a `putRequestResponse` for a future use?

Quentin01 commented on src/Application.cpp:
```diff
@@ -1,13 +1,18 @@
-#include <lug/Graphics/Vulkan/API/RTTI/Enum.hpp>
 
-Application::Application() : lug::Core::Application::Application{ {"hello", {0, 1, 0}} } {
+#include <restclient-cpp/restclient.h>
```
You don't have to include this file

Quentin01 commented on src/Application.cpp:
```diff
@@ -1,13 +1,18 @@
-Application::Application() : lug::Core::Application::Application{ {"hello", {0, 1, 0}} } {
+#include <restclient-cpp/restclient.h>
+
+#include <APIClient/Router.hpp>
```
You don't have to include this file

Quentin01 commented on src/Application.cpp:
```diff
@@ -77,296 +79,13 @@ bool Application::initDevice(lug::Graphics::Vulkan::PhysicalDeviceInfo* choosedD
+    const nlohmann::json json = GPUInfoProvider::get(*physicalDeviceInfo);
+	LUG_LOG.info("{}", json.dump());
+    APIClient::GPU::putGpu(json);
     return (true);
```
Don't put parentheses around the return value.

Quentin01 commented on src/Application.cpp:
```diff
@@ -77,296 +79,13 @@ bool Application::initDevice(lug::Graphics::Vulkan::PhysicalDeviceInfo* choosedD
-    std::cout << json.dump(4) << std::endl;
-
+    const nlohmann::json json = GPUInfoProvider::get(*physicalDeviceInfo);
+	LUG_LOG.info("{}", json.dump());
```
Use spaces for indentation

Quentin01 commented on src/GPUInfoProvider.cpp:
```diff
@@ -0,0 +1,281 @@
+        );
+    }
+
+    json["zSwapChainInfo"].push_back(
```
Rename this field `"swapchain"`

Quentin01 commented on src/GPUInfoProvider.cpp:
```diff
@@ -0,0 +1,281 @@
+            {"width", physicalDeviceInfo.swapchain.capabilities.maxImageExtent.width},
+            {"height", physicalDeviceInfo.swapchain.capabilities.maxImageExtent.height}
+        }
+        },
```
Put this bracket at the end of the previous line

Commit:
fix norm and better code overall

Fix norm in ApiClient/GPU

fix norm for common function in APIClient

router is now a namespace

vulkanInfoProvider is now a namespace

VulkanInfoProvider becomes GPUInfoProvider

move  print of json in application

remove unused http code check

add default to switch for windows

Gpu class become a namespace

there are no warning with Visual Studio 2017, so i delete useless disable 4003
